<!DOCTYPE html>
<html lang="ko">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script>
    // 3. async / await (ES2017)
    // "프로미스를 작성하기 쉽게 만듭니다."
    // async는
    // - 함수가 프로미스를 반환하도록 함
    // - 함수 앞에 붙이는 키워드
    // await는
    // - 함수가 프로미스를 기다리게 함
    // - async 함수 내에서만 사용할 수 있음
    // - 함수 실행을 일시 중지하고 프로미스 결과(성공 또는 실패)를 기다리도록 함

    // 비동기 함수 만들어 보기 예제
    // 일반적인 함수
    // TODO: 주석 풀기
    // function fetchData() {
    //   // 네트워크 요청을 통해 시간이 걸리는 작업...
    //   return new Promise((resolve, reject) => {
    //     resolve('네가 원하는 데이터');
    //   });
    // }
    // fetchData()
    //   .then(res => {
    //     console.log(res);
    //   });

    // 비동기 함수를 만들기 위해 함수 앞에 async 키워드를 붙이면 자동으로 프로미스로 반환됨 
    // TODO: 주석 풀기
    // async function fetchData2() {
    //   return 'async 네가 원하는 데이터';
    // }
    // fetchData2()
    //   .then(res => {
    //     console.log(res);
    //   });

    // // 3-1) 프로미스를 사용하는 일반적인 방식과 async / await로 리팩터링 하기
    // // 예제1
    function walk(amount) {
      return new Promise((resolve, reject) => {
        if(amount < 500) {
          reject('the value is too small');
        }
        setTimeout(() => {
          resolve(`너는 ${amount}ms만큼 걸었어!`);
        }, amount);
      });
    }
    // TODO : 주석 풀기
    walk(5000).then(res => {
      console.log(res);   //
      return walk(3000);   // walk()함수를 3초 뒤에 다시 호출
    }).then(res => {      
      console.log(res);
      return walk(1000);
    }).then(res => {
      console.log(res);   // reject에서 에러 처리 안해서 끝나는 건데 error 처리 하면 아래도 잘 나오게 할 수 있음!!
      return walk(100);
    }).then(res => {
      console.log(res);
      return walk(500);
    });


    // 예제1 리팩터링
    // TODO : 주석 풀기
    // async function go() {
    //   // 프로미스가 완료될 때까지 기다리기 위해 await 키워드를 사용
    //   // await를 쓰면 우리가 평상 시 코딩하는 것처럼 동기적 코드로 보임(가독성)
    //   const res = await walk(500);  // 0.5초 기다렸다가 결과 값을 res에 넣고 출력함(74번줄에 출력)
    //   console.log(res);
    //   const res2 = await walk(900); 
    //   console.log(res2);
    //   const res3 = await walk(600); 
    //   console.log(res3);
    //   const res4 = await walk(400); 
    //   console.log(res4);
    //   const res5 = await walk(500); 
    //   console.log(res5);
    // }
    // go();
    // await는 비동기 함수 내에서만 사용 가능!!


    // // 예제2
    // 깃허브 API로 특정 깃허브 사용자 정보 조회
    // fetch('https://api.github.com/users/leehoeeun')
    //   .then(res => {
    //     console.log(res);
    //     return res.json();
    //   })
    //   .then(res => {
    //     // 성공시 데이터를 출력
    //     console.log(res);
    //   })
    //   .catch(err => {
    //     // 실패 시 오류 출력
    //     console.error(err);
    //   });

    // // 예제2 리팩터링
    // const getGithubInfo = async username => {          // 화살표 함수가 아닌 일반 함수일 때는 async function(username){};
    //     const res = await fetch(`https://api.github.com/users/${username}`);
    //     const userInfo = await res.json();
    //     console.log(userInfo);
    //   };
    //   getGithubInfo('leehoeeun');



    // 3-2) 오류 처리
    // 일반적인 프로미스에서는 catch()를 사용하여 프로미스가 반환하는 오류들을 처리
    // async / await도 비슷 - try / catch 구문 사용
    // try에 다양한 오류가 발생한 애를 catch에서 잡아줌, 단, 오류가 다중 발생할 시 그중 첫번째만 찍힘. 하나씩 수정해 나가면 됨
    // async function asyncFunc() {
    //   try {
    //     // 에러 발생 가능한 코드들을 여기에...
    //     const response = await fetch('api-url');
    //     console.log(response);    // Response 객체, status: 404 (Not found) , ok: false
    //     if(!response.ok) {     // !response.ok = true
    //       throw new Error(`${response.status} (${response.statusText})`);     // catch에서 error를 잡을 수 있도록 오류 내용을 던짐(throw) - throw는 선택사항/근데 난 쓸래!
    //     }
    //   } catch (err) {       //  try에서 error가 발생했을 때 바로 catch문으로 간다
    //     console.error(err);
    //   }
    // }
    // asyncFunc();

    // Quiz
    // 19.1
    async function func() {
      let promise = Promise.resolve(1);
      let result = await promise;
    }
    func();



    const loginUser = (id, password) => {
      // 프로미스 객체를 만들어서 반환하고 then을 붙인다
      return new Promise((resolve, reject) => {
        setTimeout(() => {
        if(id === 'goni' && password === '1234') {
          resolve(id);   
        }else{
          reject(new Error('존재하지 않음, loginUser'));
        }
      }, 2000);
      });
    };

    const getRoles = (userId) => {
      return new Promise((resolve, reject) => {
        setTimeout(() => {
        if(userId === 'goni') {
          resolve({ name: 'leehoeeun', role: 'admin' });
        }else{
          reject(new Error('접근 권한이 없음, getRoles'));         // userId === 'goni' 에서 goni값을 변경해야 나옴
        }
      }, 1000);
      })
    };

    const id = prompt('아이디를 입력하세요.');
    const password = prompt('비밀번호를 입력하세요.');
    
    const userInfo = async userId => {
        debugger;
        const testText = await loginUser(id, password);
        const user1 = await getRoles(testText);
        const user2 = await alert(`Welcome ${user1.name} (${user1.role})`);  
        console.log(testText);
        console.log(user1);
        console.log(user2);
      }
      userInfo(id, password);





  </script>
</head>
<body>
  
</body>
</html>